core/src/main/java/org/jdbi/v3/core/argument/Arguments.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.argument;
+
+import static org.jdbi.v3.core.internal.JdbiOptionals.findFirstPresent;
+import static org.jdbi.v3.core.internal.JdbiStreams.toStream;
+
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.jdbi.v3.core.ConfigRegistry;
+import org.jdbi.v3.core.JdbiConfig;
+import org.jdbi.v3.core.array.SqlArrayArgumentFactory;
+
+public class Arguments implements JdbiConfig<Arguments> {
+
+    private final Optional<Arguments> parent;
+    private final List<ArgumentFactory> argumentFactories = new CopyOnWriteArrayList<>();
+
+    public Arguments() {
+        parent = Optional.empty();
+        register(BuiltInArgumentFactory.INSTANCE);
+        register(new SqlArrayArgumentFactory());
+    }
+
+    private Arguments(Arguments that) {
+        parent = Optional.of(that);
+    }
+
+    public Arguments register(ArgumentFactory factory) {
+        argumentFactories.add(0, factory);
+        return this;
+    }
+
+    /**
+     * Obtain an argument for given value in the given context
+     *
+     * @param type  the type of the argument.
+     * @param value the argument value.
+     * @param config the config registry, for composition
+     * @return an Argument for the given value.
+     */
+    public Optional<Argument> findFor(Type type, Object value, ConfigRegistry config) {
+        return findFirstPresent(
```

Comment:
I considered that, but I don't think we should. `ArgumentFactory.build` takes both a type and a value argument. I don't think we can assume that just because a factory returns an argument for a particular type, that it will return one for any value of that type.

But that might be a good thing to follow up on: what if Arguments were like mappers, and didn't hold values but instead received a value as a parameter and write them to a prepared statement on demand? That would definitely allow us to cache and optimize in batches.

I'll open up a separate issue to discuss this further.